### PR TITLE
Overvote

### LINKED ISSUE #260: ✨ Add a HashedElGamalCiphertext object for encryption (and decryption - currently for testing) for overvotes and writeins
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Suggestion

 An implementation of the Auxiliary Encryption as described in the ElectionGuard specification is needed to encrypt (and decrypt - currently just for testing) overvotes and writeins. 

### Possible Implementation

Create a HashedElGamalCiphertext object for encryption (and decryption - currently for testing) for overvotes and writeins. The encryption will use the Auxiliary Encryption as specified in the ElectionGuard specification.

The current padding scheme agreed on is as follows:
Two ways to encrypt.

1) If the plaintext is fixed length and a multiple of the block length (32 bytes) then the caller may indicate this and no padding is done. The code throws an exception in this case if the plaintext is not a multiple of the block length.

2) If the plaintext length is variable then a 4 byte length field is concatenated to the front of the plaintext in big endian form. This data is then padded out with 0x00 bytes to the block length. If the 4 byte length field plus the plaintext length falls on a block length boundary (for example if the plaintext is 28 bytes then the with the length concatenated then it makes it 32 bytes) then no 0x00 padding is applied. On decrypt the length is checked to make sure it is not longer than the ciphertext length minus the 4 byte length. If it is then an exception is thrown. The padding length is calculated and if any padding bytes are non 0x00 then an exception is thrown.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by jeffspel-crypto**
[//]: # (🚨 Please review the CONTRIBUTING.md in this repository. 💔Thank you!)

### Issue
*Link your PR to an issue*

Fixes #260

### Description
This PR adds the HashedElGamalCiphertext object that is to be used to encrypt/decrypt writein and overvote date. The HashedElGamalCiphertext object implements the algorithm as specified in the "Encryption of extended contest data" section of the ElectionGuard 1.01 specification.

### Testing
Multiple tests have been added for the testing the encrypt/decrypt functions. In addition a benchmark test has been added for the hashedElgamalEncrypt function.

### CODE REVIEW

File: test/electionguard/test_elgamal.cpp [hunk 1]
```diff
@@ -162,6 +164,273 @@ TEST_CASE("elgamalEncrypt encrypt 1 decrypts with secret")
...
+      move(pad), HEGResult->getData(), HEGResult->getMac());
+
+    try {
+        vector<uint8_t> new_plaintext =
+          newHEG->decrypt(*different_secret, *descriptionHash, false);
+    } catch (std::runtime_error &e) {
+        decrypt_failed = true;
+    }
+
+    CHECK(decrypt_failed);
```
[the remark this answers is not in the document]
Comment (jeffspel-crypto) [reply test/electionguard/test_elgamal.cpp:345]: Yes, good point, however the end of the test case on line 345 is a failure case where the key used for decrypt is a nonce and not the correct key. This test case is supposed to decrypt and we check to make sure it did. Make sense?

[the remark this answers is not in the document]
Comment (jeffspel-crypto) [reply test/electionguard/test_elgamal.cpp:345]: Oops mistype, that test case is supposed to NOT decrypt and we make sure it didn't.

[the remark this answers is not in the document]
Comment (SteveMaier-IRT) [reply test/electionguard/test_elgamal.cpp:345]: Thanks.  You might want to look at the others that are testing a successful case and make sure we compare values

**Review by SteveMaier-IRT — APPROVED** [review]
looks good.

File: test/electionguard/test_elgamal.cpp [hunk 2]
```diff
@@ -162,6 +164,273 @@ TEST_CASE("elgamalEncrypt encrypt 1 decrypts with secret")
...
+
+    unique_ptr<HashedElGamalCiphertext> newHEG = make_unique<HashedElGamalCiphertext>(
+      move(pad), HEGResult->getData(), HEGResult->getMac());
+
+    vector<uint8_t> new_plaintext = newHEG->decrypt(secret, *descriptionHash, false);
+
+    CHECK(plaintext == new_plaintext);
+}
+
+// it is on the boundary be
```
Comment (AddressXception) [inline test/electionguard/test_elgamal.cpp:199]: comment supposed to be here?

**Review by AddressXception — APPROVED** [review]
looks good to me - again thank you for the very descriptive unit tests

File: test/electionguard/test_elgamal.cpp (hunk 2 above)
Comment (SteveMaier-IRT) [reply test/electionguard/test_elgamal.cpp:199]: We will get this removed.

**Pull request merged**